.github/workflows/windows.yml

```diff
@@ -46,42 +46,153 @@ env:
+      selected-runner-label: ${{ steps.select.outputs.selected_runner_label }}
+      is-self-hosted: ${{ steps.select.outputs.is_self_hosted }}
+    concurrency:
+      group: servo-reserve-self-hosted-runner
+      cancel-in-progress: false
+    permissions: write-all
+    steps:
+      - name: Select and reserve best available runner
+        id: select
+        # Set the variables below to your desired runner images, runner scope
+        # (org or repo), and a token with permission to write to that scope.
+        run: |
+          github_hosted_runner_label=windows-2022
+          self_hosted_image_label=self-hosted-image:windows10
+          self_hosted_runner_scope=/orgs/${{ github.repository_owner }}/actions/runners
+          export GH_TOKEN=${{ secrets.RUNNER_API_TOKEN }}
+
+          runners=$(mktemp)
+          gh api "$self_hosted_runner_scope" > $runners
+
+          # Generate a unique id that allows the workload job to find the runner
+          # we are reserving for it (via runner labels), and allows the timeout
+          # job to find the workload job run (via the job’s friendly name), even
+          # if there are multiple instances in the workflow call tree.
+          random32=$(LC_ALL=C < /dev/urandom tr -dC 0-9a-f 2> /dev/null | head -c 8)
+          unique_id=${{ github.repository }}/${{ github.run_id }}/$random32
+          echo "unique_id=$unique_id" | tee -a $GITHUB_OUTPUT
+
+          # Find a runner that is online, not busy, and not already reserved for
+          # any job (label prefix “reserved-for:”).
+          runner_id=$(mktemp)
+          if ! < $runners > $runner_id jq \
+            --arg self_hosted_image_label "$self_hosted_image_label" -e '
+              .runners
+              | map(select(.status == "online" and .busy == false))
+              | map(select([.labels[].name] | index($self_hosted_image_label) | not | not))
+              | map(select([.labels[].name] | map(startswith("reserved-for:")) | index(true) | not))
+              | first | .id'; then
+            echo "selected_runner_label=$github_hosted_runner_label" | tee -a $GITHUB_OUTPUT
+            echo 'is_self_hosted=false' | tee -a $GITHUB_OUTPUT
+            exit 0
+          fi
+
+          # Reserve that runner by adding a label containing the unique id.
+          # Job concurrency ensures that runners never get assigned twice.
+          reserved_since=$(date +\%s)
+          gh api "$self_hosted_runner_scope/$(cat $runner_id)/labels" \
+            -f "labels[]=reserved-for:$unique_id" \
+            -f "labels[]=reserved-since:$reserved_since" \
+            --method POST --silent
+          echo "selected_runner_label=reserved-for:$unique_id" | tee -a $GITHUB_OUTPUT
+          echo 'is_self_hosted=true' | tee -a $GITHUB_OUTPUT
+
+  # In the unlikely event the selected runner goes down after we reserve it but
+  # before the workload starts, cancel the workflow run.
+  runner-timeout:
```

Comment:
runner-timeout must only run for self-hosted runner, as it's normal to wait a long time for github runner if there is a lot of runs (in my personal profile for example). `if: ${{ fromJSON(needs.runner-select.outputs.is-self-hosted) }}` on job (not step) should do the trick.

Reply:
done in ff5ad6f51f9f756143713b06121856626e09e6ee ([build](https://github.com/delan/servo/actions/runs/10505379092/job/29102818331), [build](https://github.com/delan/servo/actions/runs/10505381412/job/29102826536))

Reply:
one sec, the build failed because it ran on an old runner